crates/wit-component/tests/components/imports/component.wat

```diff
@@ -0,0 +1,166 @@
+  (module (;1;)
+    (type (;0;) (func (param i32 i32)))
+    (type (;1;) (func (param i32)))
+    (type (;2;) (func))
+    (type (;3;) (func (param f32)))
+    (func (;0;) (type 0) (param i32 i32)
+      local.get 0
+      local.get 1
+      i32.const 0
+      call_indirect (type 0)
+    )
+    (func (;1;) (type 1) (param i32)
+      local.get 0
+      i32.const 1
+      call_indirect (type 1)
+    )
+    (func (;2;) (type 0) (param i32 i32)
+      local.get 0
+      local.get 1
+      i32.const 2
+      call_indirect (type 0)
+    )
+    (func (;3;) (type 2)
+      i32.const 3
+      call_indirect (type 2)
+    )
+    (func (;4;) (type 1) (param i32)
+      local.get 0
+      i32.const 4
+      call_indirect (type 1)
+    )
+    (func (;5;) (type 2)
+      i32.const 5
+      call_indirect (type 2)
+    )
+    (func (;6;) (type 1) (param i32)
+      local.get 0
+      i32.const 6
+      call_indirect (type 1)
+    )
+    (func (;7;) (type 3) (param f32)
+      local.get 0
+      i32.const 7
+      call_indirect (type 3)
+    )
+    (table (;0;) 8 8 funcref)
+    (export "bar1" (func 0))
+    (export "bar2" (func 1))
+    (export "baz1" (func 2))
+    (export "baz2" (func 3))
+    (export "baz3" (func 4))
+    (export "foo1" (func 5))
+    (export "foo2" (func 6))
+    (export "foo3" (func 7))
+    (export "$imports" (table 0))
+  )
+  (module (;2;)
+    (type (;0;) (func (param i32 i32)))
+    (type (;1;) (func (param i32)))
+    (type (;2;) (func))
+    (type (;3;) (func (param f32)))
+    (import "" "bar1" (func (;0;) (type 0)))
+    (import "" "bar2" (func (;1;) (type 1)))
+    (import "" "baz1" (func (;2;) (type 0)))
+    (import "" "baz2" (func (;3;) (type 2)))
+    (import "" "baz3" (func (;4;) (type 1)))
+    (import "" "foo1" (func (;5;) (type 2)))
+    (import "" "foo2" (func (;6;) (type 1)))
+    (import "" "foo3" (func (;7;) (type 3)))
+    (import "" "$imports" (table (;0;) 8 8 funcref))
+    (elem (;0;) (i32.const 0) func 0 1 2 3 4 5 6 7)
+  )
+  (instance (;3;) (instantiate (module 1)))
+  (instance (;4;) (instantiate (module 0) (with "bar" (instance 3)) (with "baz" (instance 3)) (with "foo" (instance 3))))
+  (alias export (instance 3) "$imports" (table (;0;)))
+  (alias export (instance 0) "bar1" (func (;0;)))
+  (alias export (instance 0) "bar2" (func (;1;)))
+  (alias export (instance 1) "baz1" (func (;2;)))
+  (alias export (instance 1) "baz2" (func (;3;)))
+  (alias export (instance 1) "baz3" (func (;4;)))
+  (alias export (instance 2) "foo1" (func (;5;)))
+  (alias export (instance 2) "foo2" (func (;6;)))
+  (alias export (instance 2) "foo3" (func (;7;)))
+  (func (;8;) (canon.lower utf8 (into (instance 4)) (func 0)))
+  (func (;9;) (canon.lower (func 1)))
+  (func (;10;) (canon.lower utf8 (into (instance 4)) (func 2)))
+  (func (;11;) (canon.lower (func 3)))
+  (func (;12;) (canon.lower (func 4)))
+  (func (;13;) (canon.lower (func 5)))
+  (func (;14;) (canon.lower (func 6)))
+  (func (;15;) (canon.lower (func 7)))
```

Comment:
Oh, that's an _excellent_ point that I completely looked over since originally I was needlessly slapping `into` on every lowering with the expectation that it'd get smarter (it did, but then I didn't update the import shimming to take this into account).

I'm happy to fix this prior to merging.

Reply:
I've pushed up a new commit that only indirectly lowers imports that need it; if not, they get lowered before the core module is instantiated and passed directly as an argument.